Add context.go and context_test.go

Riomhchlaraitheoir opened the pull request:
These files were copied from benbjohnson/clock with needed modifications

This allows creating context with timeout/deadline based on the clock

Commit:
Add context.go and context_test.go

These files were copied from benbjohnson/clock with
needed modifications

jfrabaute commented on context.go:
```diff
@@ -0,0 +1,86 @@
+	"time"
+)
+
+func ContextWithTimeout(parent context.Context, clock Clock, timeout time.Duration) (context.Context, context.CancelFunc) {
```
Could we enforce ContextWithTimeoutCause and ContextWithDeadlineCause (Meaning with don't have ContextWithTimeout and ContextWithDeadline at all as an option), so we can have a cause, that might help with debugging when we track context cancellations.

jfrabaute replied:
CC @aaronbee because I'd like to enforce this, but maybe we still want to give people options?

Commit:
Add ContextWithDeadlineCause function

ContextWithDeadlineCause and ContextWithTimeoutCause can be used to add a cause to the context timeout

Riomhchlaraitheoir commented:
I've added them in a second commit, it was a bit more difficult than I had expected since only a `context.WithCancelCause` context can have a cause, but I'm confident that it's now working as intended

Riomhchlaraitheoir commented:
I also left the functions without causes, since that's already the pattern in the context package, while it might be nice to enforce it here, IMO that's not as useful when the context package doesn't do this. I'll defer to your decision on this though
